pinot-spi/src/main/java/org/apache/pinot/spi/env/PinotConfiguration.java

```diff
@@ -170,8 +174,15 @@ public static List<Configuration> applyDynamicEnvConfig(List<Configuration> conf
       Map<String, String> environmentVariables) {
     return configurations.stream().peek(configuration -> {
       for (String dynamicEnvConfigVarName : configuration.getStringArray(ENV_DYNAMIC_CONFIG_KEY)) {
-        configuration.setProperty(dynamicEnvConfigVarName,
-            environmentVariables.get(configuration.getString(dynamicEnvConfigVarName)));
+        if (configuration.getProperty(TEMPLATED_KEY) == null || !configuration.getProperty(TEMPLATED_KEY)
+            .equals(true)) {
+          configuration.setProperty(dynamicEnvConfigVarName,
+              environmentVariables.get(configuration.getString(dynamicEnvConfigVarName)));
+          //Make sure the env variable is not re read twice by setting the property of Templated = true
+          configuration.setProperty(TEMPLATED_KEY, true);
```

Comment:
We cannot set it here because it will prevent other configs to be set. You may add the checkout outside of the for loop